src/callgraphwidget.cpp

```diff
@@ -0,0 +1,208 @@
+    setLayout(layout);
+
+    KService::Ptr service = KService::serviceByDesktopName(QStringLiteral("kgraphviewer_part"));
+
+    if (!service) {
+        return;
+    }
+    auto costThreshold = new QDoubleSpinBox(this);
+    costThreshold->setDecimals(2);
+    costThreshold->setMinimum(0);
+    costThreshold->setMaximum(99.90);
+    costThreshold->setPrefix(tr("Cost Threshold: "));
+    costThreshold->setValue(m_threshold);
+    costThreshold->setSingleStep(0.1);
+    costThreshold->setSuffix(QStringLiteral("%"));
+
+    connect(costThreshold, QOverload<double>::of(&QDoubleSpinBox::valueChanged), this, [this](double threshold) {
+        m_threshold = threshold;
+        generateCallgraph(m_currentSymbol);
+    });
+
+    m_graphview = service->createInstance<KParts::ReadOnlyPart>(this);
+    if (!m_graphview) {
+        return;
+    }
+
+    layout->addWidget(costThreshold);
+    layout->addWidget(m_graphview->widget());
+
+    m_interface = qobject_cast<KGraphViewer::KGraphViewerInterface*>(m_graphview);
+
+    updateColors();
+    m_interface->setLayoutMethod(KGraphViewer::KGraphViewerInterface::LayoutMethod::InternalLibrary);
+
+    connect(m_graphview, SIGNAL(hoverEnter(QString)), this, SLOT(hoverEnter(QString)));
+    connect(m_graphview, SIGNAL(hoverLeave(QString)), this, SLOT(hoverLeave(QString)));
+
+    // used to find mouseclicks in the kgraphviewer part
+    qApp->installEventFilter(this);
+
+    m_graphFile->open();
+
+    connect(Settings::instance(), &Settings::callgraphChanged, this, [this]{
+        generateCallgraph(m_currentSymbol);
+    });
+}
+
+void CallgraphWidget::selectSymbol(const Data::Symbol& symbol)
+{
+    if (symbol == m_currentSymbol) {
+        return;
+    }
+
+    generateCallgraph(symbol);
+}
+
+void CallgraphWidget::hoverEnter(const QString& node)
+{
+    m_currentNode = node;
+}
+
+void CallgraphWidget::hoverLeave(const QString& node)
+{
+    Q_UNUSED(node);
+    m_currentNode = QString();
+}
+
+bool CallgraphWidget::eventFilter(QObject* watched, QEvent* event)
+{
+    Q_UNUSED(watched);
+    if (!isVisible())
+        return false;
+
+    if (event->type() == QEvent::MouseButtonPress) {
+        const auto e = static_cast<QMouseEvent*>(event);
+
+        if (m_graphview->widget()->geometry().contains(e->pos())) {
+            if (e->button() == Qt::MouseButton::LeftButton && !m_currentNode.isEmpty()) {
+                if (m_nodeIdToIndex.contains(m_currentNode)) {
+                    emit callerCalleeIndexChanged(m_nodeIdToIndex[m_currentNode]);
+                    m_currentNode.clear();
+                }
+            }
+        }
+    }
+    return false;
+}
+
+void CallgraphWidget::changeEvent(QEvent* event)
+{
+    QWidget::changeEvent(event);
+
+    if (event->type() == QEvent::PaletteChange) {
+        updateColors();
+    }
+}
+
+void CallgraphWidget::showEvent(QShowEvent* event)
+{
+    if (!m_graphview) {
+        return;
+    }
+
+    if (!event->spontaneous()) {
```

Comment:
can you please document why you are checking for spontaneous events here? first time I'm seeing this

Reply:
spntaneous means that the event was fired after the widget became visible.
seems that this is not needed, I just read https://doc.qt.io/qt-5/qwidget.html#showEvent and thought that this was required.